jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryEndpointCaller.java

```diff
@@ -125,7 +130,38 @@
    */
   @Nullable
   T call() throws IOException, RegistryException {
-    return call(initialRequestUrl);
+    return callWithAllowInsecureRegistryHandling(initialRequestUrl);
+  }
+
+  @Nullable
+  private T callWithAllowInsecureRegistryHandling(URL url) throws IOException, RegistryException {
+    try {
+      if (!isHttpsProtocol(url) && !allowInsecureRegistries) {
+        throw new InsecureRegistryException(url);
+      }
+      return call(url, connectionFactory);
+
+    } catch (SSLPeerUnverifiedException exception) {
+      try {
+        // Cannot verify the server (e.g., self-signed certificate or plain HTTP).
+        if (!allowInsecureRegistries) {
+          throw new InsecureRegistryException(url);
+        }
+        if (insecureConnectionFactory == null) {
+          insecureConnectionFactory = Connection.getInsecureConnectionFactory();
+        }
+        return call(url, insecureConnectionFactory);
+
+      } catch (GeneralSecurityException ex) {
+        throw new RegistryException("cannot turn off TLS peer verification", ex);
+
+      } catch (SSLPeerUnverifiedException | HttpHostConnectException ex) {
+        // Try HTTP as a last resort.
```

Comment:
For example, here, I think this would allow HTTP fallback even if `allowInsecureRegistries` was false.

Reply:
That can never happen. If `allowInsecureRegistries` is false and the server is not a verifiable HTTPS, it will throw an exception.

```java
        // Cannot verify the server (e.g., self-signed certificate or plain HTTP).
        if (!allowInsecureRegistries) {
          throw new InsecureRegistryException(url);
        }
```

Reply:
Ah right, I need to read this more carefully.